Enable saving/exporting changelog data

Achoobert opened the pull request:
This needs to be merged in order for the Changelog object to work in the ABdesigner

I'm not sure this is the right branch?

Achoobert committed:
[add] keep version number as created elsewhere

https://github.com/digi-serve/plugin_ABDesigner/issues/108

Achoobert committed:
[fix] move versionData to it's own object in `json`

Also change it to an Object format rather than just a string

Additionally, if data does not exist: create an "initial" version with 'initial' changelog

Achoobert committed:
match style expected in designer

Hiro-Nakamura commented on ABDefinitionCore.js:
```diff
@@ -42,6 +42,15 @@ module.exports = class ABDefinitionCore {
       this.type = attributes.type || attributes?.json?.type || "";
       this.json = attributes.json || null;
+
+      this.version = attributes.version || attributes.json?.version || "1.0.0";
```
I don't think we need to store this information at the `ABDefinition` level.  

Currently versioning is for `ABApplication`s and you already have it being pulled out and defined there.   Adding it in here will attempt to pull out version info for EVERYTHING in our AppBuilder definitions: ABObject, ABFields, ABQueries, ABDatacollections, tasks, etc...  

So let's just leave it at the `ABApplication` level.

Hiro-Nakamura reviewed (changes requested):
Modify the ABDefinition code.

Achoobert committed:
[fix] data structure

https://github.com/digi-serve/plugin_ABDesigner/issues/153

Achoobert committed:
fix default datastructure

Don't set the versionNumber at the application level

Achoobert committed:
fix: setting data in wrong order

Hiro-Nakamura commented on ABApplicationCore.js:
```diff
@@ -66,6 +66,24 @@ module.exports = class ABApplicationCore extends ABMLClass {
+         },
+      };
+
+      if (typeof this.versionData == "string")
```
I don't see where `this.versionData` was set previously.  

Is this supposed to be `this.json.versionData` ?

Hiro-Nakamura commented on ABApplicationCore.js:
```diff
@@ -278,6 +296,10 @@ module.exports = class ABApplicationCore extends ABMLClass {
 
       this.json.name = this.name;
 
+      this.json.versionNumber = this.versionData.versionNumber;
```
we seem to be specifically storing the `.versionNumber` in `this.json` but we don't seem to be reading it out or accessing it as `this.json.versionNumber` anywhere.  

I think it would be just as good to just reference it as `this.versionData.versionNumber`.

So we probably don't need to store it like this.

Hiro-Nakamura commented on ABDefinitionCore.js:
```diff
@@ -42,6 +42,17 @@ module.exports = class ABDefinitionCore {
       this.type = attributes.type || attributes?.json?.type || "";
       this.json = attributes.json || null;
+
+      this.json.versionData = attributes.json?.versionData || {
```
This is what you are already doing in `ABApplicationCore`.  We don't need to make this assignment here as well.

Achoobert committed:
[fix] requested refactor

Code refactor done, duplicates we bet,
Data setup and versionNumber unset.
Oh dear me, please do tell,
What's the purpose of obj to string spell?

Achoobert committed:
[fix] stop language updating deleting the versionData

Achoobert commented:
Ok, I think I isolated everything version-related down to the .json level. 

I had to put a little hack into the toObj() function, as returning that was overwriting the whole .json and wiping out the versionData. The result is a much cleaner export of the data as well

Hiro-Nakamura commented on ABDefinitionCore.js:
```diff
@@ -42,6 +42,8 @@ module.exports = class ABDefinitionCore {
       this.type = attributes.type || attributes?.json?.type || "";
       this.json = attributes.json || null;
+
+      this.json.versionData = attributes.json?.versionData;
```
don't put this here.

Hiro-Nakamura commented on ABMLClassCore.js:
```diff
@@ -96,6 +96,7 @@ module.exports = class ABMLClassCore extends ABEmitter {
 
       return {
          translations: this.translations,
+         versionData: this.json?.versionData,
```
or here

Hiro-Nakamura reviewed (changes requested):
Yeah, this `ABApplicationCore` object is being a little tricky with it's `.json` attribute.  

Instead of putting the `versionData` in the lower `ABDefinition` or `ABMLObject` methods, keep it in the `ABApplicationCore.toObj()`.  You just need to grab the value out before the  
```
this.json = super.toObj();
```
call and then tuck it back in later. 

Something like:
```
let versionData = this.json.versionData;
this.json = super.toObj();

this.json.name = this.name;
this.json.versionData = versionData;
...
```

Achoobert committed:
[fix] keep versionData in ABApplicationCore and extract before toObj() 

In lower ABDefinition and ABMLObject methods,
VersionData brought many bugs and mods,
Now in ABApplicationCore.toObj() it shall stay,
Just grab the value out before .toObj() takes it away.

Hiro-Nakamura reviewed (approved):
great!
